Feat/Scrum 167 노트 신고, 삭제, 수정 관련 UI 및 로직 구현 + (유튜브 뮤직 실행 기능)

inwoodev opened the pull request:
## PR 타입
어떤 변경에 대한 PR인가요?

<!-- 아래 체크리스트 중 해당 되는 부분에 체크해주세요 "x". -->

- [ ] 버그 수정
- [x] Feature / 신기능
- [ ] 코드 스타일 업데이트 (formatting, local variables)
- [ ] 리팩토링 (no functional changes, no api changes)
- [ ] 빌드 관련 변경사항
- [ ] CI 관련된 변경사항
- [ ] Documentation content changes
- [ ] 구조 변경
- [ ] Other... 추가 설명 요망:


## 배경
기획 요구사항에 맞게 게시된 노트에 추가 기능이 필요합니다

유저 자신이 게시한 노트인 경우
- 노트 수정
- 노트 삭제

다른 유저가 게시한 노트인 경우
- 노트 신고

## 목표

<!-- 해당 PR에서 개발/수정 한 기능의 목표를 작성해주세요. -->
- [x] 노트 미트볼 메뉴 구현
- [x] 노트 삭제 기능 구현(작성자인경우)
- [x] 음악 플레이버튼 탭할 경우 유튜브 뮤직으로 이동하는 로직 구현 

추후 구현
- 노트 수정화면으로 화면전환(추후 coordinator 활용 예정)
- 노트 신고화면으로 화면전환(추후 coordinator 활용 예정)

## 결과 (Optional)

### 노트 삭제, 신고, 수정 미트볼 메뉴:
![Simulator Screen Recording - iPhone 15 Pro - 2024-09-02 at 01 38 13](https://github.com/user-attachments/assets/a5b79543-50f8-4946-bfb5-a256c27f1be9)

---

|수정/삭제 | 신고|
|:-----:|:-----:|
| ![GIF 1](https://github.com/user-attachments/assets/8b7ee864-2faa-4b68-863d-3a79306bb213) | ![GIF 2](https://github.com/user-attachments/assets/3c864bee-cc10-4cf4-b152-d40a0bfff3dc) |

---

### 플레이버튼 탭 이후 유튜브 뮤직 실행:
1. 유튜브 뮤직 앱이 있을 경우 앱 실행 및 관련 노래 검색
2. 유튜브 뮤직 앱이 없을 경우 유튜브 뮤직을 브라우저를 통해 실행


| YouTubeMusic 有 | YouTubeMusic 無 |
|:-----:|:-----:|
| <img src="https://github.com/user-attachments/assets/7af2d5fa-19de-4bbb-b43f-3e01bff27f9a" width="350"/> | <img src="https://github.com/user-attachments/assets/b524a3c1-97e4-41a6-9610-3ff98073b2e8" width="350"/> |

inwoodev committed:
add: 노트의 미트볼 메뉴 관련 에셋 추가

inwoodev committed:
feat: 노트의 미트볼 메뉴 UI 구현

그외:
- UIColor Extension 접근제한 수정(internal → public)
- BottomSheetViewController의 view를 외부에서 주입 받을 수 있도록 로직 수정

inwoodev committed:
refactor: 토큰 관련 response 모델 수정

- 사용되지 않는 모델 제거
  - ReissueTokenResponse
  - UserValidityResponse

- 토큰 모델 수정
  - 네이밍 수정: TokenResponse → UserAuthResponse
  - userId프로퍼티 추가

inwoodev committed:
refactor: userID대신 userInformation을 저장하도록 로직 수정

- 저장될 UserInformation 관련 키 네이밍 변경: USER_ID_KEY → USER_INFO_KEY
- UserInformation모델 구현

inwoodev committed:
feat, test: 회원가입시 전달받는 userId를 저장하는 로직 추가 & 유닛테스트 추가

inwoodev committed:
move: UserInformation 파일 이동

- CoreLocalStorage모듈에서 SharedUtil모듈로 이동
(VC, ViewModel, Coordinator등에서 사용될 모델이기에)

inwoodev committed:
feat: 노트의 미트볼 메뉴 비즈니스 로직 추가

- 서버로부터 받은 유저아이디와 메칭되는 노트인 경우:
  - 수정 버튼, 삭제 버튼 화면 출력
- 유저아이디와 매칭되지 않는 노트인 경우:
  - 신고 버튼 화면 출력

inwoodev committed:
feat: 노트 삭제 기능 구현

- note id를 활용하여 삭제 api 콜 호출
- delete api콜 성공시 관련 노트 삭제
- delete api콜 실패시 노트 삭제 x

inwoodev committed:
refactor : 노트의 미트볼 메뉴의 높이 설정 로직 버그 수정

- 180 고정 값이 아닌 노트타입에 따라 높이가 변경되도록 수정

inwoodev committed:
fix: 작성자 비교 로직 수정

- userID와 note id를 비교하는 것이 아니라 publisher.id와 비교해야 한다.

inwoodev commented on Plugins/DependencyPlugin/ProjectDescriptionHelpers/Project+Environment.swift:
```diff
@@ -128,7 +128,7 @@ public extension Project {
                 ],
                 "ACCESS_TOKEN_KEY": "${ACCESS_TOKEN_KEY}",
                 "REFRESH_TOKEN_KEY": "${REFRESH_TOKEN_KEY}",
-                "USER_ID_KEY": "${USER_ID_KEY}"
+                "USER_INFO_KEY": "${USER_INFO_KEY}"
```
단순히 UserID만 저장할 수 있는게 아니라 유저 관련된 조금 더 범용성 있는 모델로 만들어서 저장하는게 낫다고 판단이 되었기에 UserInformation을 저장하는 방향으로 수정을 하였고 모델 변경에 따라서 사용되는 키 네이밍 또한 위처럼 수정하였습니다.

@derrickkim0109 TokenKeys.xcconfig 수정 하시면 좋을 듯 합니다.

derrickkim0109 replied:
넵 알겠습니다~!

inwoodev commented on Projects/Core/Network/Interface/Sources/DTO/UserLoginResponse.swift:
```diff
@@ -7,7 +7,8 @@
 import Foundation
 
-public struct TokenResponse: Decodable {
+public struct UserAuthResponse: Decodable {
     public let accessToken: String
     public let refreshToken: String
+    public let userId: Int
 }
```
userId가 추가 되었기에 TokenResponse보다는 조금 더 범용적인 네이밍이 더 나을듯 하여 위처럼 수정하였습니다. 어떤가요~?

derrickkim0109 replied:
네 좋습니다~!

inwoodev commented on Projects/Core/LocalStorage/Sources/Storage/UserInfoStorage.swift:
UserInformation을 어떤식으로 저장할까 고민하다 TokenStorage와 비슷하게 구현하면 적어도 저장하는 부분은 테스트할 수 있다는 생각에 유사하게 구현 해 보았습니다.

물론 ViewController 및 Feature 모듈에서는 KeychainWrapper를 사용하셔도 무방합니다!

inwoodev commented on Projects/Feature/Main/Project.swift:
```diff
@@ -54,5 +54,8 @@ let targets: [Target] = [
 let project = Project.makeModule(
     name: ModulePath.Feature.name + ModulePath.Feature.Main.rawValue,
-    targets: targets
+    targets: targets,
+    additionalFiles: [
+        .folderReference(path: .relativeToRoot("./Projects/App/xcconfigs"))
+    ]
```
Example앱에서 root의 xcconfig폴더에 접근할 수 없어 TokenKey를 가져오지 못하는 문제를 위와 같이 .folderReference를 지정하여 해결하였습니다 :)

inwoodev commented on Projects/Feature/Main/Interface/Sources/ViewControllers/MainViewController.swift:
```diff
@@ -16,6 +16,10 @@ public class MainViewController: UIViewController {
     
     private var cancellables: Set<AnyCancellable> = .init()
     
+    // MARK: - Keychain
+    @KeychainWrapper<UserInformation>(.userInfo)
+    var userInfo
```
UserInformation이 Shared모듈에 있기에 편하게 사용할 수 있습니다 :)

derrickkim0109 replied:
너무 좋습니다~!

inwoodev commented on Projects/Shared/Util/Sources/Extensions/Combine/UIViewController+Combine.swift:
```diff
@@ -0,0 +1,20 @@
+//
+//  Created by 황인우 on 9/2/24.
+//
+
+import Combine
+import UIKit
+
+public extension UIViewController {
+    func dismissPublisher(animated: Bool) -> AnyPublisher<Void, Never> {
+        return Future { [weak self] promise in
+            self?.dismiss(animated: animated, completion: {
+                promise(.success(()))
+            })
+        }
+        .eraseToAnyPublisher()
+    }
+}
```
viewController가 dismiss 완료 되는 시점에 이벤트를 방출하는 publisher를 구현 해 봤습니다:)

derrickkim0109 replied:
오 저도 바로 사용하면 좋을 것 같군요~!

inwoodev commented on Projects/Feature/Main/Interface/Sources/ViewControllers/NoteMenuViewConroller.swift:
```diff
@@ -11,10 +11,66 @@ import Combine
+        onReport: PassthroughSubject<Int, Never>,
+        onEdit: PassthroughSubject<Int, Never>,
+        onDelete: PassthroughSubject<Int, Never>
+    ) {
```
해당 뷰컨을 띄우는 MainViewController의 passsthrough subject들에게 신고, 수정, 삭제가 되는 시점을 알려줄 수 있게 위와 같이 subject들을 주입하는 방향으로 구현하였습니다~!

inwoodev commented on Projects/Feature/Main/Interface/Sources/ViewModels/MainViewModel.swift:
```diff
@@ -226,3 +227,22 @@ extension MainViewModel {
     }
 }
+
+// MARK: - Edit, Delete, Report Note
+extension MainViewModel {
+    func deleteNote(id: Int) {
+        self.deleteNoteUseCase.execute(noteID: id)
+            .mapToResult()
+            .receive(on: DispatchQueue.main)
+            .sink { result in
+                switch result {
+                case .success:
+                    self.fetchedNotes.removeAll(where: { $0.id == id })
+                    
+                case .failure(let noteError):
+                    self.error = .noteError(noteError)
+                }
+            }
+            .store(in: &cancellables)
+    }
+}
```
삭제 api콜 성공시에만 일단은 해당되는 노트를 삭제하도록 로직을 구현하였습니다~!

inwoodev committed:
refactor: 노트 삭제 팝업 문구 수정

inwoodev committed:
feat: 음악 플레이버튼을 탭할 때 유튜브 뮤직앱으로 이동하는 로직 추가

derrickkim0109 reviewed (approved):
수고하셨습니다~!

inwoodev committed:
refactor : 홈화면과 관련된 이미지 에셋 추가 및 교체

- 1x, 2x, 3x에 적합하지 않은 이미지 에셋이 추가되어서 이미지가 깨져보이는 문제 수정
- 2x, 3x 에셋이 존재하지 않은 경우 이미지 추가
- 중복되는 에셋 제거

inwoodev committed:
refactor: Main → Home으로 네이밍 수정

- 홈화면을 담당하는 객체들이기에 Main이라는 키워드 보다 Home키워드가 더 적합하다고 판단 후 수정
수정 내역:
  - MainViewController → HomeViewController
  - MainViewModel → HomeViewModel
  - MainView → HomeView
